### PR TITLE
Port broccoli config loader

### PR DESCRIPTION
**Pull request by yaymukund**
This addresses the first part of [Extract Broccoli Plugins (ember-cli/issues/4907)](https://github.com/ember-cli/ember-cli/issues/4907) See https://github.com/ember-cli/ember-cli/pull/4976 for the ember-cli integration.

### CODE REVIEW

**Comment by trabus** [comment]
@chadhietala @stefanpenner @rwjblue If any of you have time, please review. I'll merge and publish after a :+1:

File: index.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+function fuzzyPurgeRequireEntry(entry) {
+  return Object.keys(require.cache).filter(function(path) {
+    return path.indexOf(entry) > -1;
+  }).forEach(function(entry) {
+    delete require.cache[entry];
+  });
+}
+
+ConfigLoader.prototype.clearConfigGeneratorCache = function() {
+  fuzzyPurgeRequireEntry(this.options.project.configPath()+'.js');
```
Comment (chadhietala) [inline index.js]: spacing between `+`

File: index.js [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+  if (this.options.tests) {
+    environments.push('test');
+  }
+
+  environments.forEach(function(env) {
+    var config = this.options.project.config(env),
+        jsonString = JSON.stringify(config),
+        outputPath = path.join(outputDir, env);
+
+    fs.writeFileSync(outputPath+'.json', jsonString, {
```
Comment (chadhietala) [inline index.js]: Same as above.

**Comment by yaymukund** [comment]
Thanks for reviewing, @chadhietala !

File: package.json [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    "ember-cli",
+    "javascript",
+    "js"
+  ],
+  "author": "yaymukund <yaymukund+broccoli-babel-transpiler@gmail.com>",
+  "license": "MIT",
+  "bugs": {
+    "url": "https://github.com/babel/broccoli-config-loader/issues"
+  },
+  "homepage": "https://github.com/babel/broccoli-config-loader#readme",
```
Comment (kellyselden) [inline package.json]: the refs to babel in this file can be moved over to ember-cli

File: test/fixtures/project.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+var Project = function(config) {
+  this.root = tmp.in(tmpRoot);
+  var configDir = join(this.root, 'config');
+  fs.mkdirSync(configDir);
+  this.writeConfig(config);
+};
+
+Project.prototype.writeConfig = function(config) {
+  var configPath = this.absoluteConfigPath(),
+      contents = 'module.exports = function() { return '+JSON.stringify(config)+'; };';
```
Comment (kellyselden) [inline test/fixtures/project.js]: "+" spacing

File: test/fixtures/project.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+  fs.writeFileSync(configPath+'.js', contents, { encoding: 'utf8' });
```
Comment (kellyselden) [inline test/fixtures/project.js]: same as above

**Comment by kellyselden** [comment]
@yaymukund Added some comments, then a squash would be great!

**Comment by yaymukund** [comment]
Okay, I grepped for `"\S+\S"` and I think you caught all of them. I'll be more mindful of that in the future. Thanks, @kellyselden

**Comment by kellyselden** [comment]
LGTM

**Pull request merged**